xonsh/events.py

```diff
@@ -0,0 +1,169 @@
+"""
+Events for xonsh.
+
+In all likelihood, you want builtins.__xonsh_events__
+
+The best way to "declare" an event is something like::
+
+    __xonsh_events__.on_spam.doc("Comes with eggs")
+"""
+import abc
+import collections.abc
+import traceback
+import sys
+
+
+class AbstractEvent(collections.abc.MutableSet, abc.ABC):
+    def __call__(self, handler):
+        """
+        Registers a handler. It's suggested to use this as a decorator.
+
+        A decorator method is added to the handler, validator(). If a validator
+        function is added, it can filter if the handler will be considered. The
+        validator takes the same arguments as the handler. If it returns False,
+        the handler will not called or considered, as if it was not registered
+        at all.
+        """
+        #  Using Pythons "private" munging to minimize hypothetical collisions
+        handler.__validator = None
+        self.add(handler)
+
+        def validator(vfunc):
+            """
+            Adds a validator function to a handler to limit when it is considered.
+            """
+            handler.__validator = vfunc
+        handler.validator = validator
+
+        return handler
+
+    def _filterhandlers(self, handlers, *pargs, **kwargs):
+        """
+        Helper method for implementing classes. Generates the handlers that pass validation.
+        """
+        for handler in handlers:
+            if handler.__validator is not None and not handler.__validator(*pargs, **kwargs):
+                continue
+            yield handler
+
+    @abc.abstractmethod
+    def fire(self, *pargs, **kwargs):
+        """
+        Fires an event, calling registered handlers with the given arguments.
+        """
+
+
+class Event(AbstractEvent):
+    """
+    A given event that handlers can register against.
+
+    Acts as a ``set`` for registered handlers.
+
+    Note that ordering is never guaranteed.
+    """
+    # Wish I could just pull from set...
+    def __init__(self):
+        self._handlers = set()
+
+    def __len__(self):
+        return len(self._handlers)
+
+    def __contains__(self, item):
+        return item in self._handlers
+
+    def __iter__(self):
+        yield from self._handlers
+
+    def add(self, item):
+        self._handlers.add(item)
+
+    def discard(self, item):
+        self._handlers.discard(item)
+
+    def fire(self, *pargs, **kwargs):
+        """
+        Fires each event, returning a non-unique iterable of the results.
+        """
+        vals = []
+        for handler in self._filterhandlers(self._handlers, *pargs, **kwargs):
+            try:
+                rv = handler(*pargs, **kwargs)
+            except Exception:
+                print("Exception raised in event handler; ignored.", file=sys.stderr)
+                traceback.print_exc()
```

Comment:
This should probably use xonsh.tools.print_traceback() https://github.com/xonsh/xonsh/blob/master/xonsh/tools.py#L590 so that it respects the user configuration settings.